### PR TITLE
feat: Let the database generate UUIDs on insert

### PR DESCRIPTION
**Pull request by alnr**
Combined with https://github.com/gobuffalo/pop/pull/807, this will allow for fast-path inserts for tables with UUID as primary key.

### CODE REVIEW

File: continuity/container.go [hunk 1]
```diff
@@ -48,7 +48,7 @@ func (c Container) TableName(ctx context.Context) string {
 
 func NewContainer(name string, o managerOptions) *Container {
 	return &Container{
-		ID:         x.NewUUID(),
+		ID:         uuid.Nil,
```
Comment (jonas-jonas) [inline continuity/container.go:51]:
```suggestion
```

File: persistence/sql/persister_session.go [hunk 2]
```diff
@@ -180,9 +180,13 @@ func (p *Persister) UpsertSession(ctx context.Context, s *session.Session) (err
 	s.NID = p.NetworkID(ctx)
 
 	return errors.WithStack(p.Transaction(ctx, func(ctx context.Context, tx *pop.Connection) error {
-		exists, err := tx.Where("id = ? AND nid = ?", s.ID, s.NID).Exists(new(session.Session))
-		if err != nil {
-			return sqlcon.HandleError(err)
+		exists := false
+		if !s.ID.IsNil() {
+			var err error
```
Comment (jonas-jonas) [inline persistence/sql/persister_session.go:185]:
```suggestion
```

Right?

File: persistence/sql/persister_errorx.go [hunk 3]
```diff
@@ -35,7 +33,7 @@ func (p *Persister) Add(ctx context.Context, csrfToken string, errs error) (uuid
 	}
 
 	c := &errorx.ErrorContainer{
-		ID:        x.NewUUID(),
+		ID:        uuid.Nil,
```
Comment (jonas-jonas) [inline persistence/sql/persister_errorx.go:36]:
```suggestion
```

**Review by jonas-jonas — CHANGES_REQUESTED** [review]
Just a few nits/cleanup opportunities (while we are at it ;)). Very nice cleanup! I like it! 🎉

**Comment by alnr** [comment]
Regarding the `ID: uuid.Nil` comment on struct initialization: I wanted to make it clear that, when creating a new object, you don't get a valid UUID until you've persisted it. This changes the semantics somewhat, and is not easily possible in some places in Kratos. Still working on it.

File: examples/go/selfservice/recovery/main_test.go [hunk 4]
```diff
@@ -19,7 +19,7 @@ func TestFunc(t *testing.T) {
 	publicURL, _ := testhelpers.StartE2EServer(t, "../../pkg/stub/kratos.yaml", nil)
 	client = pkg.NewSDKForSelfHosted(publicURL)
 
-	flow := performRecovery("dev+" + uuid.New().String() + "@ory.sh")
+	flow := performRecovery("dev+" + uuid.Must(uuid.NewV4()).String() + "@ory.sh")
```
[the remark this answers is not in the document]
Comment (alnr) [reply examples/go/selfservice/recovery/main_test.go:22]: This leads (I think) to import cycles. I'd also prefer to make the examples not depend on kratos internals

File: examples/go/selfservice/verification/main_test.go [hunk 5]
```diff
@@ -19,7 +19,7 @@ func TestFunc(t *testing.T) {
 	publicURL, _ := testhelpers.StartE2EServer(t, "../../pkg/stub/kratos.yaml", nil)
 	client = pkg.NewSDKForSelfHosted(publicURL)
 
-	flow := performVerification("dev+" + uuid.New().String() + "@ory.sh")
+	flow := performVerification("dev+" + uuid.Must(uuid.NewV4()).String() + "@ory.sh")
```
[the remark this answers is not in the document]
Comment (alnr) [reply examples/go/selfservice/verification/main_test.go:22]: see above

File: driver/registry_default.go [hunk 6]
```diff
@@ -586,71 +586,77 @@ func (m *RegistryDefault) Init(ctx context.Context, ctxer contextx.Contextualize
 	bc := backoff.NewExponentialBackOff()
 	bc.MaxElapsedTime = time.Minute * 5
 	bc.Reset()
-	return errors.WithStack(
```
Comment (alnr) [inline driver/registry_default.go:589]: This change is much smaller if ignoring whitespace.

File: persistence/sql/persister_session.go (hunk 2 above)
Comment (alnr) [reply persistence/sql/persister_session.go:185]: Doesn't matter, really.

File: session/manager_http.go [hunk 7]
```diff
@@ -261,23 +267,28 @@ func (s *ManagerHTTP) DoesSessionSatisfy(r *http.Request, sess *Session, request
...
+		i := sess.Identity
+		if i == nil {
+			i, err = s.r.IdentityPool().GetIdentity(ctx, sess.IdentityID, identity.ExpandCredentials)
+			if err != nil {
+				return err
+			}
+			sess.Identity = i
+		} else if len(i.Credentials) == 0 {
+			// If credentials are not expanded, we load them here.
+			if err := s.r.PrivilegedIdentityPool().HydrateIdentityAssociations(r.Context(), i, identity.ExpandCredentials); err != nil {
```
Comment (zepatrik) [inline session/manager_http.go:279]:
```suggestion
			if err := s.r.PrivilegedIdentityPool().HydrateIdentityAssociations(ctx, i, identity.ExpandCredentials); err != nil {
```

File: driver/registry.go [hunk 8]
```diff
@@ -197,6 +199,12 @@ func ReplaceTracer(f func(*otelx.Tracer) *otelx.Tracer) func(o *options) {
 	}
 }
 
+func Inspect(f func(reg Registry) error) func(o *options) {
```
Comment (zepatrik) [inline driver/registry.go:202]: What is this used for? Just external usage?

**Review by zepatrik — COMMENTED** [review]
Nice, I just have a few questions but no blocking issue.

File: identity/identity.go [hunk 9]
```diff
@@ -301,7 +299,10 @@ func (i *Identity) ParseCredentials(t CredentialsType, config interface{}) (*Cre
 }
 
 func (i *Identity) CopyWithoutCredentials() *Identity {
-	var ii = *i
+	i.lock().RLock()
```
[the remark this answers is not in the document]
Comment (alnr) [reply identity/identity.go:302]: correct

File: driver/registry.go (hunk 8 above)
Comment (alnr) [reply driver/registry.go:202]: Exactly.

File: continuity/test/persistence.go [hunk 10]
```diff
@@ -62,6 +64,7 @@ func TestPersister(ctx context.Context, p interface {
 			expected := createContainer(t)
 
 			require.NoError(t, p.SaveContinuitySession(ctx, &expected))
+			require.NotEqual(t, uuid.Nil, expected.ID)
```
[the remark this answers is not in the document]
Comment (alnr) [reply continuity/test/persistence.go:67]: I couldn't find an existing reference to `require.Zero` or `require.NotZero` in the codebase, so I kept with the existing style. We could refactor all instances at one in a later PR if we want to.

File: driver/registry_default.go [hunk 11]
```diff
@@ -586,71 +586,77 @@ func (m *RegistryDefault) Init(ctx context.Context, ctxer contextx.Contextualize
...
+		m.persister = p.WithNetworkID(net.ID)
+		return nil
+	}, bc)
 
-			m.persister = p.WithNetworkID(net.ID)
-			return nil
-		}, bc),
-	)
+	if err != nil {
+		return errors.WithStack(err)
```
[the remark this answers is not in the document]
Comment (alnr) [reply driver/registry_default.go]: fixed

File: session/manager_http.go (hunk 7 above)
Comment (alnr) [reply session/manager_http.go]: Good catch, fixed.

File: driver/registry_default.go [hunk 12]
```diff
@@ -586,71 +586,77 @@ func (m *RegistryDefault) Init(ctx context.Context, ctxer contextx.Contextualize
...
 
-			m.persister = p.WithNetworkID(net.ID)
-			return nil
-		}, bc),
-	)
+	if err != nil {
+		return err
+	}
+
+	if o.inspect != nil {
```
Comment (zepatrik) [inline driver/registry_default.go:654]: Hm not sure what the use-case is, but I see two code-paths that `return nil` before this is reached. Just FYI.

**Review by zepatrik — APPROVED** [review]
LGTM :+1:

File: session/manager_http_test.go [hunk 13]
```diff
@@ -549,6 +549,25 @@ func TestDoesSessionSatisfy(t *testing.T) {
 			} else {
 				require.NoError(t, err)
 			}
+
+			// This should still work even if the session does not have identity data attached yet...
```
Comment (aeneasr) [inline session/manager_http_test.go:553]: Why? :D Or rather, why was it not an issue before?

**Review by aeneasr — APPROVED** [review]
LGTM! Once comments from @zepatrik are addressed this is mergable :)

File: session/manager_http_test.go (hunk 13 above)
Comment (alnr) [reply session/manager_http_test.go:553]: This change makes our code more resilient against future changes with regards to expanding or not expanding the identity, for example during `whoami` calls. Previously, we would rely on the `whoami` handler to know that it needed the identity credentials later, and "pre-load" them. That's brittle. I played around with trying to enable a whoami call without expanding the identity (`GET `/sessions/whoami?expand=none`, for performance) and the handler would panic.

This change allows us to more aggressively minimize necessary queries without breaking functionality.

File: driver/registry_default.go (hunk 12 above)
Comment (aeneasr) [reply driver/registry_default.go:654]: @alnr please double check this

Comment (alnr) [reply driver/registry_default.go:654]: You mean on lines 637 and 647? Those returns are returning from the closure passed to `backoff.Retry`. We return from `Init` only if `backoff.Retry` returns non-nil. So this codepath is always reached in the happy path.

**Comment by aeneasr** [comment]
Nice, LGTM. Rerunning CI will hopefully resolve the curl timeouts.

**Pull request merged**